minimint/src/lib.rs

```diff
@@ -145,6 +173,89 @@ impl MinimintServer {
+        let mut msg_buffer: Vec<PeerMessage> = vec![];
+        let mut next_epochs = BTreeMap::<PeerId, u64>::new();
+
+        tracing::info!("Rejoining consensus");
+        while next_epochs.len() < self.cfg.max_faulty() + 1 {
+            match self.connections.receive().await {
+                (peer, EpochMessage::RejoinEpoch(epoch)) => {
+                    next_epochs.insert(peer, epoch);
+                }
+                msg if next_epochs.contains_key(&msg.0) => msg_buffer.push(msg),
+                (_, _) => {}
+            };
+        }
+        let next_epoch_num = next_epochs.into_values().max().unwrap();
+        self.hbbft.skip_to_epoch(next_epoch_num);
+        tracing::info!("Rejoining consensus: at epoch {}", next_epoch_num);
+
+        // Run consensus until arriving at the next_epoch consensus outcome
+        let mut outcomes: Vec<ConsensusOutcome> = vec![];
+        let mut next_outcome = None;
+
+        for msg in msg_buffer {
+            outcomes.append(&mut self.handle_message(msg).await);
+        }
+
+        while next_outcome.is_none() {
+            let msg = self.connections.receive().await;
+            outcomes.append(&mut self.handle_message(msg).await);
+            next_outcome = outcomes.iter().find(|out| out.epoch == next_epoch_num);
+        }
+        tracing::info!("Rejoining consensus: created outcome");
+
+        // Download epochs and verify hashes
+        let mut epochs: Vec<EpochHistory> = vec![];
+        let saved_epoch_key = self.consensus.db.get_value(&LastEpochKey).unwrap();
+        let download_epoch_num = saved_epoch_key.clone().map(|e| e.0 + 1).unwrap_or(0);
+        let mut prev_epoch = saved_epoch_key.and_then(|e| self.consensus.db.get_value(&e).unwrap());
+
+        for i in download_epoch_num..next_epoch_num {
+            let epoch = self.api.fetch_epoch_history(i).await.unwrap();
+            assert!(epoch.verify_hash(&prev_epoch).is_ok());
+            prev_epoch = Some(epoch.clone());
+            epochs.push(epoch);
+            tracing::info!("Rejoining consensus: downloaded epoch {}", i);
+        }
+
+        // Add the outcome we just got from consensus
+        let contributions = next_outcome.unwrap().contributions.clone();
+        let end_epoch = EpochHistory::new(next_epoch_num, contributions, &prev_epoch);
+        epochs.push(end_epoch.clone());
+
+        // Verify signature if we downloaded any epochs
+        if let Some(mut prev_epoch) = prev_epoch {
+            assert!(end_epoch
+                .add_sig_to_prev(&self.cfg.epoch_pk_set, &mut prev_epoch)
+                .is_ok());
+        }
+
+        // process and save epochs
+        for epoch in epochs {
+            tracing::info!(
+                "Rejoining consensus: processing epoch {}",
+                epoch.outcome.epoch
+            );
+            let outcome = ConsensusOutcomeConversion::from(epoch.outcome).0;
+            self.consensus.process_consensus_outcome(outcome).await;
```

Comment:
Do we need a struct for the conversion here? Wouldn't a direct impl of `From` work? Or just a plain old function?

Reply:
Once we can get remove `ConsensusOutcomes` dependency on hbbft then we can eliminate this.